Dual F9P for yaw with proper redundancy behaviour

oystub opened the pull request:
**NB! Due to time constraints, this text is not fully up to date!**

## Desired end state

The goal is to use two uBlox F9P GNSS receivers in a way where:
- Under normal operation, the receivers should provide heading using RTK in a moving base/rover configuration
- If a single receiver malfunctions, the autopilot should still get accurate position information, even though heading is no longer available. Malfunctions mean:
   - Signal conditions too bad to calculate an accurate and stable navigation solution, this can be caused by local interference or bad connection to the antenna.
   - The receiver loses fix, but still sends messages to the autopilot. For example if vibrations/temperature fluctuations/firmware issues causes the unit to lose fix or reset some internal state
   - No valid messages are received by the autopilot. Causes include power loss, signal interference, broken signal cable.

## Problem 1: uBlox module behaviour in rover mode causes timing issues

When two uBlox receivers are connected in a moving base/rover pair, if the rover is unable to calculate a navigation solution, it will reduce the navigation rate all the way to 1 Hz. This is not a huge problem in itself, however, the problem is that the content of the messages are then also ~900 ms delayed. That is, both the `time_utc_usec` field and the actual position data in the GNSS message is 900 ms old when received by PX4. Even if compensated for, allowing for such a long delay would require significantly increasing the EKF  fusion time horizon, with all associated disadvantages.

<img width="1280" height="117" alt="image" src="https://github.com/user-attachments/assets/aa7fb923-8a82-4d50-9172-69c31da4019f" />
Note that this behavior is described in the integration manual and isn’t possible to disable.

PX4 doesn’t use the time_utc_usec field, it just considers the system time when it received the message, and subtracts a constant delay as configured in EKF2_GPS_DELAY. It will therefore treat the 900 ms messages as if they are only ~100 ms old, and try to fuse them into the position estimate. We have observed this “delayed mode” if the signal conditions are poor (something we experience regularly on the ground) or if the moving base receiver drops out. In multiple events, this has caused severe navigation issues and toilet bowl/local position loss, for example in https://aviant.atlassian.net/wiki/spaces/TECHNICAL/pages/1312063493.

If the messages from the moving base unit completely stops, the rover remains in this mode for `CFG-NAVSPG-CONSTR_DGNSSTO` seconds, and then continues as normal. However, if it receives intermittent data, it will go back into the problematic mode again, potentially remaining there for an indefinite period of time.

Example from testing on the ground. The moving base unit (gps 01) is plugged out, causing the rover unit (gps 00) to experience reduced frequency and delayed content in GPS messages. `CFG-NAVSPG-CONSTR_DGNSSTO=60` (default).

<img width="1280" height="525" alt="image" src="https://github.com/user-attachments/assets/a7ffd88d-de88-4116-a789-519976c26f8b" />
Example from testing on the ground. The moving base unit (gps 01) is plugged out, causing the rover unit (gps 00) to experience reduced frequency and delayed content in GPS messages. `CFG-NAVSPG-CONSTR_DGNSSTO=60` (default).

## Problem 2: Bad handling of GNSS heading and blending in PX4

PX4 comes with a feature called “GPS blending”. When using blending, sensor messages from both connected GNSS receivers are fused into one “combined” measurement with weights based on the reported accuracy of the messages. The blending code also estimates the position difference between the two units, so that the blended messages represent the position of the midpoint between the receivers. There is currently no way of specifying body-frame offsets for different GPS receivers.

Currently, the only way to use the heading estimate from the rover, is to have blending enabled. In this case, if the rover provides a value in the heading field of the `sensor_gps` message, it is used in the blended output. 

Because of the aforementioned timing issues that can happen with rover GNSS receivers, we cannot use blending, since the timestamps of the messages from the rover cannot be trusted. Therefore, there is no way to use the heading measurements from the rover without also relying on it for the global absolute position.

Some additional background: uBlox GNSS receivers (ubx version 7 and newer) in rover mode actually transmit the global navigation solution, and the relative position to the moving base (including heading angle), in two different messages, `UBX_NAV_PVT` and `UBX_NAV_RELPOSNED`, respectively. The PX4 uBlox diver combine these into a single sensor_gps uorb message, with the heading fields originating from the `RELPOSNED` message. However, since https://github.com/PX4/PX4-GPSDrivers/pull/101, the relative position information is additionally published directly to the `sensor_gnss_relative` uorb topic for each received `RELPOSNED` message . However, as of now, these messages are not used by any other modules.

Even though not a direct issue, in my opinion, absolute GNSS position and GNSS heading should be treated separately, instead of being combined in this hacky way in both the driver and the blending module. This would significantly clean up the architecture. The relative measurements, including heading, are inherently a different thing from the absolute position measurements.

## Why is this a problem for us?

See section above. The GPS data from the rover GPS will have wrong timestamp if the signal conditions are bad or the moving base drops out, which again causes navigation issues and/or failsafes => We don’t have proper redundancy for the moving base GPS

As long as we don’t have a reliable time delay for the rover GPS, we cannot navigate use the measurements from the rovers, not even in GPS blending. Because of this, we currently have no way of using the heading measurements from the rover in a safe way. 

I.e., we dont have a way to use the RTK heading during normal operation, while still having reliable redundancy if the MB GPS drops out or experiences bad signal conditions.

## Solution
The development consists of two main “modules”

1. Dynamically re-configuring the rover unit to avoid time delay problems
2. Using gnss heading measurements independently of blending.

> Note: In an ideal world, I would prefer to always use “GNSS” instead of “GPS” and “heading” instead of “yaw”, since they are most technically correct in this case. (Yaw  only equals heading because of the chosen Euler rotation sequence).

### Dynamic reconfiguration

> This solution is only for receivers connected over serial. DroneCAN receivers would need their own firmware updates and DroneCAN commands to do this.

#### Reset procedure

After testing multiple ways of resetting the rover unit, the following process was found to be most reliable and caused the shortest GPS downtime:

- Disable RTCM input on the port from the moving base
- Enable RTCM input on the port from the flight controller, so reference station corrections can be used
- Disable `NAV_RELPOSNED` message output
- Perform a GPS-only software reset of the position
- In testing, this procedure regains fix within 1-2 seconds from issuing the command 

- [ ] TODO: Insert test results

A function has been implemented in the ublox driver for issuing these commands. 

There are multiple potential ways of triggering the reset procedure.

In this implementation, each sensor instance of the sensor/gps module subscribes to the `vehicle_gps_position` topic. If all the following conditions are met

- GPS1 (which is always rover) is selected or used in blending. NB! In later PX4 version, we use device_id instead of selected field.
- This instance is configured as rover
- Some configurable time has elapsed since boot (to allow receivers some time to come online and get fix)

then the rover unit is permanently (until next reboot) reset to operate as a normal receiver (not in rover mode). A warning message is displayed to the operator.

### Better GNSS yaw handling
A new sensor bridge module is introduced, which publishes `vehicle_gps_yaw`. It serves the same role as `vehicle_gps_position`, but takes `sensor_gnss_relative` as input, and outputs to the `vehicle_gps_yaw` uorb topic.

Like other sensor bus modules, it has the task of calibration and validation of the incoming raw sensor data. For simplicity in the initial design, the new module does not handle selection. That is, it doesn’t support multiple rovers. However, since PX4 currently doesn’t support more than two GNSS receivers anyway, this is not deemed a big problem. The module also translates the raw baseline heading angle into the heading of the body frame relative to NED, based on the specified antenna position.

#### EKF2 modifications
The remaining implementation is changing the EKF2 to use the new sensor bridge module instead of vehicle_gps_position for heading fusion.

## Before merge

- [x] Merge https://github.com/aviant-tech/PX4-GPSDrivers/pull/1 and update submodule pointer

bresch committed:
gps_blending: fix selection rapid switching

Once a timeout of the primary instance is detected, a fallback is only
allowed until the primary receiver is regained.

cherry-pick of upstream 094048e

bresch committed:
gps_blending: fallback to secondary if primary has no fix

cherry-pick of upstream b8c81f6

oystub committed:
cubeorange: (aviant) disable some unused modules to free up flash

oystub committed:
gps: point to Aviant fork of device driver submodule

oystub commented on src/modules/sensors/vehicle_gnss_relative/params.c:
```diff
@@ -0,0 +1,60 @@
+ * GPS Yaw Baseline Length Accuracy Requirement
+ *
+ * Discard any GPS yaw data if the baseline length deviates from the expected length by more than this value.
+ * Expected length is calculated from the SENS_GPSYAW_P{X,Y,Z} parameters.
```
```suggestion
 * Expected length is calculated from the SENS_GPSREL_P{X,Y,Z} parameters.
```

Torjusba commented on src/modules/sensors/vehicle_gnss_relative/VehicleGPSYaw.cpp:
```diff
@@ -0,0 +1,168 @@
+		_parameter_update_sub.copy(&param_update);
+
+		updateParams();
+		_sensor_gnss_relative_sub.registerCallback();
```
It makes more sense for me to put this in the Start function, since the registration only needs to happen once, and it does not depend on any parameter values.

oystub replied:
It's based on the vehicle_gps_position code, ideally i want to be able to update params live. I'll check the performance hit and consider doing reboot-only parameter load.

oystub replied:
TODO: vehicle_gps_position maybe does it a bit differently

oystub replied:
Nvm, I read your comment a bit too quickly, and thought you meant the parameter update check. 

Of course callback registration should only be called once!

Torjusba commented on src/modules/sensors/vehicle_gnss_relative/VehicleGPSYaw.cpp:
```diff
@@ -0,0 +1,168 @@
+		// Rotate into vehicle frame using configured antenna position
+		const float bx = _params_gpsrel_px.get();
+		const float by = _params_gpsrel_py.get();
+		const float baseline_body_angle = atan2f(by, bx);
```
Getting the parameter values both here and in `updateParameters` looks a bit unnecessary to me, but I'm not sure if it's measurably slower in practice, I guess it's a pretty cheap operation.

oystub replied:
Yeah, I guess i need to look a bit more into the param handling! I don't think `get` actually fetches the updated param value. It just returns the current stored value. So both calls are needed.

oystub replied:
removed the trig call

Torjusba commented on src/modules/sensors/vehicle_gnss_relative/VehicleGPSYaw.cpp:
```diff
@@ -0,0 +1,168 @@
+	perf_begin(_cycle_perf);
+	ParametersUpdate();
+
+	if (!isnanf(_expected_baseline_length_m) && _sensor_gnss_relative_sub.updated()) {
```
Any specific reason for using `ìsnanf` instead of `PX4_ISFINITE` here?

oystub replied:
No, just a slip up

Torjusba commented on src/modules/sensors/vehicle_gnss_relative/VehicleGPSYaw.cpp:
```diff
@@ -0,0 +1,168 @@
+			.heading_accuracy = d.heading_accuracy
+		};
+
+		_vehicle_gps_yaw_pub.publish(out);
```
Consider adding a perf counter of type `PC_COUNT` here, to log the count of published messages from this module. Then we can more easily separate the cases "didn't receive heading" and "discarded heading" during troubleshooting.
(unless we log both `vehicle_gps_yaw` and `sensor_gnss_relative` at a high enough rate, then we will already have that info in our log)

Torjusba commented on src/modules/ekf2/EKF/gps_yaw_fusion.cpp:
```diff
@@ -53,16 +53,12 @@ void Ekf::fuseGpsYaw()
-	const Vector3f ant_vec_bf = {cosf(_gps_yaw_offset), sinf(_gps_yaw_offset), 0.0f};
-	const Vector3f ant_vec_ef = _R_to_earth * ant_vec_bf;
+	// First column of rotation matrix = body x-axis in earth frame
+	const Vector3f ant_vec_ef = _R_to_earth.col(0);
```
I don't think it' safe to assume that the antenna array will always be inline with the body x axis.

Especially since the introduction of separate `SENS_GPSREL_(XYZ)` parameters indicates that the implementation supports arbitrary relative antenna locations.

oystub replied:
`vehicle_gps_yaw` is already rotated into body frame. But i I should change the variable naming or add a comment here so that this is easy to understand.

oystub replied:
TODO: find out why. and why isnt roll checked?

Torjusba commented on src/modules/sensors/vehicle_gnss_relative/VehicleGPSYaw.hpp:
```diff
@@ -0,0 +1,55 @@
+
+namespace sensors
+{
+class VehicleGPSYaw : public ModuleParams, public px4::ScheduledWorkItem
```
IMO this class and file name should be the same as the folder name, keeping to the existing convention in the sensors directory. E.g. `vehicle_gps_position` contains `VehicleGPSPosition`

oystub replied:
Yeah, my thought was that there could be other uses of relative gnss than just heading. But that might be a bit of a stretch.

oystub replied:
I.e. collect all of the "sensor_gnss_relative" based modules in that directory

Torjusba reviewed (changes requested):
I like the overall approach and implementation, left a few comments and suggestions.

> In an ideal world, I would prefer to always use “GNSS” instead of “GPS” and “heading” instead of “yaw”, since they are most technically correct in this case. (Yaw only equals heading because of the chosen Euler rotation sequence).

I agree, not sure if it's better to stick with `vehicle_gps_yaw` in order to maintain similarity to the existing implementation or go for `vehicle_gnss_heading` for correctness.

oystub commented:
> I like the overall approach and implementation, left a few comments and suggestions.
> 
> > In an ideal world, I would prefer to always use “GNSS” instead of “GPS” and “heading” instead of “yaw”, since they are most technically correct in this case. (Yaw only equals heading because of the chosen Euler rotation sequence).
> 
> I agree, not sure if it's better to stick with `vehicle_gps_yaw` in order to maintain similarity to the existing implementation or go for `vehicle_gnss_heading` for correctness.

I have thought a bit about it, and i prefer just going full in for gnss and heading. Seems like this is the sentiment in discussions upstream as well.

Torjusba reviewed (approved):
lgtm

oystub committed:
sensors: add vehicle_gnss_heading to sensor bridge

Modelled after vehicle_gps_position, it processes incoming sensor_gnss_relative
messages and outputs vehicle_gps_yaw with a validated body-frame heading
measurement.

In the future, this can be used in the EKF, instead of the existing hacky
solution which only works with blending enabled.

Note that only one instance is supported. This is because PX4 only has support
for two GNSS receivers anyway, so there isn't "room" for multiple rovers that
output heading.

oystub committed:
ekf2: use vehicle_gnss_heading as source for rtk heading measurements

oystub committed:
gps: option for normal mode fallback if ubx mb rover is used for position

A receiver in moving base rover mode will output data with delay up to 1 second
if it cannot compute a relative solution. This can for example happen if the
moving base experiences degrated signal or power loss.

In such a scenario, the rover might be the only receiver left, and for
redundancy it's important to bring it back into a normal state without delay in
order to avoid toilet bowl/position loss.
